### PR TITLE
junbon/volsurface_validator

### PR DESCRIPTION
**Pull request by junbon**
This PR is based off junbon/volsurface_refactoring

### CODE REVIEW

File: lib/Quant/Framework/VolSurface.pm [hunk 1]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+        try { $self->$method } catch { $self->validation_error($_) };
+    }
+
+    return !$self->validation_error;
+}
+
+sub _validate_age {
+    my $self = shift;
+
+    if (time - $self->recorded_date->epoch > 4 * 3600) {
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: What if there is a `for_date`?

File: lib/Quant/Framework/VolSurface.pm [hunk 2]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+    # Somehow I do not know why there is a limit of term on delta surface, but
+    # for moneyness we might need at least up to 2 years to get the spread.
+    my $type = $self->type;
+
+    my $extra_allowed = $self->underlying_config->extra_vol_diff_by_delta || 0;
+    my $max_vol_change_by_delta = 0.4 + $extra_allowed;
+
+    my @days = sort { $a <=> $b } keys %{$surface_hashref};
+
+    if (@days < 2) {
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: It will be more readable to use `scalar` or even @# notation

File: lib/Quant/Framework/VolSurface.pm [hunk 3]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+        die('Must be at least two maturities on vol surface.');
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: Include symbol name in the validation error.

File: lib/Quant/Framework/VolSurface.pm [hunk 4]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+
+        for (my $i = 0; $i < $#volatility_level; $i++) {
+            my $level = $volatility_level[$i];
+
+            if ($level !~ /^\d+\.?\d+$/) {
+                die("Invalid vol_point[$level] for underlying[$system_symbol].");
+            }
 
-    return !$err;
+            my $next_level = $volatility_level[$i + 1];
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: Will this work when `$i` is `$#volatility_level-1`?

File: lib/Quant/Framework/VolSurface.pm [hunk 5]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+    }
+
+    return;
+}
+
+sub _is_valid_volatility_smile {
+    my ($self, $smile) = @_;
+
+    foreach my $vol (values %$smile) {
+        return if ($vol !~ /^\d?\.?\d*$/ or $vol > 5);
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: Please add a comment here explaining reason for 5 and the regex.

File: lib/Quant/Framework/VolSurface.pm [hunk 6]
```diff
@@ -684,29 +685,190 @@ sub fetch_historical_surface_date {
...
+    foreach my $term (@existing_terms) {
+        my $existing_smile = $existing_surface_data->{$term}->{smile};
+        my $new_smile      = $new_surface_data->{$term}->{smile};
+        return if (scalar(keys %$existing_smile) != scalar(keys %$new_smile));
+        return if (any { $existing_smile->{$_} != $new_smile->{$_} } keys %$existing_smile);
+    }
+
+    my $existing_surface_epoch = Date::Utility->new($self->document->{date})->epoch;
+
+    if (time - $existing_surface_epoch > 15000 and not $self->underlying_config->quanto_only) {
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface.pm]: What is we have `for_date`?

File: lib/Quant/Framework/VolSurface/Moneyness.pm [hunk 7]
```diff
@@ -385,6 +374,125 @@ sub _convert_moneyness_smile_to_delta {
...
+                moneyness => ($vol_level + 2),
+                %vol_period,
+            });
+            my $bet_plus_h = Math::Business::BlackScholes::Binaries::vanilla_call($S, $barrier + $h, $t, $r, $r - $q, $vol);
+
+            ## The actual finite difference formula has a / $h**2, but since we're checking
+            ## for negativity, we don't need it. Also, we introducted an error margin to allow
+            ## surfaces that are close to passing through.
+            my $convexity_flag = $bet_minus_h - 2 * $bet + $bet_plus_h;
+            if ($convexity_flag < -0.009) {
```
Comment (mm-binary) [inline lib/Quant/Framework/VolSurface/Moneyness.pm]: Why 0.009?

File: t/VolSurface/validator.t [hunk 8]
```diff
@@ -181,13 +178,12 @@ subtest 'Unit test tools.' => sub {
             chronicle_writer  => $chronicle_w
         });
 
-    lives_ok { $validator->validate_surface($sample_surface) } 'Our default sample surface is valid.';
+    lives_ok { $sample_surface->is_valid } 'Our default sample surface is valid.';
```
Comment (mm-binary) [inline t/VolSurface/validator.t]: Seems that `is_valid` never dies, but just sets `validation_error`

File: lib/Quant/Framework/VolSurface.pm (hunk 1 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: validation will only be called when we save a new volsurface, not during
pricing. So for_date does not matter here.

File: lib/Quant/Framework/VolSurface.pm (hunk 2 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: Perl will do the right thing. I think this is fine.

File: lib/Quant/Framework/VolSurface.pm (hunk 3 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: noted.

File: lib/Quant/Framework/VolSurface.pm (hunk 6 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: same as the above. ```for_date``` does not apply here.

File: lib/Quant/Framework/VolSurface.pm (hunk 5 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: noted

File: lib/Quant/Framework/VolSurface.pm (hunk 4 above)
Comment (junbon) [inline lib/Quant/Framework/VolSurface.pm]: ```$#volatility_level``` return the maximum index of an array. So this will work

**Pull request merged**